### PR TITLE
Make autoprefixed tree configurable

### PR DESCRIPTION
**Pull request by jschilli**
I'll put this up for discussion - it does depend on ember-cli@0.2.0 (for running tests cleanly)

In short, I need this to prevent the 'all' tree from rerunning autoprefixer.

We have a huge project and once ember serve is running a simple change to a js file ends up running the autoprefixer even though no css tree has been invalidated.

I don't know the motivation for including all hence i made it an option with the current behavior as a default.

This would be considerably cleaner but less flexible if we just autoprefixed the 'styles' tree

awesome plugin

### CODE REVIEW

**Comment by kimroen** [comment]
If you check out [the source](https://github.com/ember-cli/ember-cli/blob/v0.2.0/lib/broccoli/ember-app.js#L321) for `addonPostprocessTree`, which is where `postprocessTree` on the addons are called and then look around for where that function is used, you will see that it is never called with `'styles'` as a parameter.

**Comment by kimroen** [comment]
I was actually a bit unsure when I checked out this part of the code to review this - why do I look for `'app'` in the first place, and not just `'styles'`?

It turns out that if you check out [the source](https://github.com/ember-cli/ember-cli/blob/v0.2.0/lib/broccoli/ember-app.js#L321) for `addonPostprocessTree` – which is where `postprocessTree` on the addons are called – and then look around for where that function is used, you will see that it is never called with `'styles'` as a parameter. The css is always delivered as a part of the `all`-tree.

So yeah - turns out there's no `postprocessTree` for styles specifically, which is probably something we should look in to adding to ember-cli. When that is done, I see no reason to run autoprefixer on the `all`-tree in addition to the `styles`-tree.

File: index.js [hunk 1]
```diff
@@ -2,23 +2,34 @@
...
-      enabled: true
-    });
-    this.enabled = this.options.enabled;
-    delete this.options.enabled;
-  },
-  postprocessTree: function(type, tree) {
-    if ((type === 'all' || type === 'styles') && this.enabled) {
-      tree = autoprefixer(tree, this.options);
-    }
+    name: 'ember-cli-autoprefixer',
```
Comment (kimroen) [inline index.js]: Looks like you accidentally changed the indentation of the code in this file to be four spaces instead of two.

File: index.js [hunk 2]
```diff
@@ -2,23 +2,34 @@
...
+    init: function() {
+        this.options = null;
+    },
+    included: function(app, parentAddon) {
+        this._super.included(app);
+        this.app = app;
+		// Only process app options the first time
```
Comment (kimroen) [inline index.js]: This shouldn't be necessary, I'm pretty sure `included` is only called once when the addon is being included in the running of the app.

**Comment by kimroen** [comment]
That said, there might be some way to misuse some of the other hooks to figure out if the styles tree changed and process `all` only if that was the case.

File: index.js (hunk 2 above)
Comment (jschilli) [inline index.js]: I added it in direct response to observing that it was called twice.

In our app (as facilitated by ember-cli) we have multiple trees - the 2nd time it's called the options.enabled is `deleted` so bombs out

File: index.js (hunk 1 above)
Comment (jschilli) [inline index.js]: sorry about that - I can clean up and resubmit

Comment (jschilli) [inline index.js]: indeed - confirmed that it is not called with 'style' - long night evidently here

I will review alternate approaches in the coming week

Comment (kimroen) [inline index.js]: Like I said in my comment, I think adding it to ember-cli is probably the way to go.

**Comment by jschilli** [comment]
added issue to ember-cli https://github.com/ember-cli/ember-cli/issues/3534 to cover this.

will look into it more next week

**Comment by kimroen** [comment]
Thanks! Really appreciate you taking the time to look in to this.

**Comment by jschilli** [comment]
@kimroen https://github.com/ember-cli/ember-cli/issues/3534 landed - I'll rework the fix to look for 'css' as a tree.

I would expect that we'll still need to do a set of trees to watch so that we can preserve compatibility with ember-cli prior to the release that this drops into - at some point (1.0?) we can strip that out

will try to put up a PR in the next day or so

**Comment by kimroen** [comment]
@jschilli I see there's been a lot of back and forth on this with changes and fixes in ember-cli. Would you mind summarizing what the status is?

**Comment by jschilli** [comment]
the fix to call  `postProcessTree` with 'css' as type landed.  `ember-cli-autoprefixer` should now be able to listen to this iso `all`

if you're willing to rev the version number and break compatibility prior to ember-cli 0.2.2 I can put up a PR that only listens for 'css'.  If you'd prefer a backwardly compatible change I can reintroduce the `autoprefixTrees` logic to allow that to be overridden

**Comment by kimroen** [comment]
That's what I thought, but I saw Stefan mentioning that there was a bigger issue that would prevent it from working like we want.

In either case, I think I'd rather check for the ember-cli version number and check for `'all'` in previous versions, and `'css'` in newer ones. Using `ember-cli-version-checker` [like we did in `ember-cli-coffeescript`](https://github.com/kimroen/ember-cli-coffeescript/commit/90b5f352f3682cab79fe95b2952820e5e06ccd5d) is probably the way to go.

Thanks for the update!

**Comment by kandhavivekraj** [comment]
@jschilli  @kimroen any update on this?

**Pull request closed without merging**